Watchman.Discord/Areas/Users/Controllers/UsersController.cs

```diff
@@ -44,10 +48,50 @@ public void GetAvatar(DiscordRequest request, Contexts contexts)
         [DiscordCommand("add role")] //todo
         public void AddRole(DiscordRequest request, Contexts contexts)
         {
-            var commandRole = request.OriginalMessage.ToLowerInvariant().Replace("-add role ", string.Empty); //TODO use DiscordRequest properties
-            var safeRoles = this._queryBus.Execute(new GetDiscordServerSafeRolesQuery(contexts.Server.Id)).SafeRoles;
-            var messagesService = _messagesServiceFactory.Create(contexts);
-            _rolesService.AddRoleToUser(safeRoles, messagesService, contexts, commandRole);
+            var args = request.Arguments.Skip(1).ToArray();
+            var rolesToAssign = args.Select(x => x.Value).ToList();
+
+            if (rolesToAssign.Count() < 1)
+            {
+                throw new NotEnoughArgumentsException();
+            }
+
+            if (args.GroupBy(x => x.Value)
+                .Select(x => x.First()).Count() != args.Length)
+            {
+                throw new ArgsAreDuplicatedException();
+            }
+
+            var allRoleNames = _usersRolesService.GetRoles(contexts.Server).ToList().Select(x => x.Name);
+            var safeRoleNames = _queryBus.Execute(new GetDiscordServerSafeRolesQuery(contexts.Server.Id)).SafeRoles.ToList().Select(x => x.Name);
+            var userRoles = contexts.User.Roles.Select(x => x.Name);
```

Comment:
myśle że ta logika powinna być w serwisie

za dużo pchasz do kontrolera
kontroler powinien tylko z góry sterować przepływem danych
a u ciebie jest masa operacji niskopoziomowych (w kontekście tego co ma się stać)

według dobrych praktyk metody kontrolerów nie powinny przekraczać 50 linii kodu => 50 to z założenia jest potwór z masą logiki pod spodem 
u ciebie metoda AddRole() ma coś w okolicy 50 linii kodu
mimo że z poziomu logiki aplikacji, to proste dodawanie wielu roli na raz, a nie bardzo skomplikowana wieloetapowa akcja z masą zależności pod spodem 

podstawowy błąd początkujących programistów => strach przed tworzeniem wielu klas i rozdzielaniem kodu między nie